modules/microvm/virtualization/microvm/host/storage.nix

```diff
@@ -0,0 +1,115 @@
+# Copyright 2022-2024 TII (SSRC) and the Ghaf contributors
+# SPDX-License-Identifier: Apache-2.0
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}: let
+  cfg = config.ghaf.services.storage;
+  inherit (builtins) filter map listToAttrs;
+  inherit (lib) mkIf optionals optionalAttrs optionalString forEach nameValuePair recursiveUpdate concatStringsSep unique foldl';
+
+  # Helper functions
+  filterHostShares = filter (s: s.tag != null && s.host-path != null) cfg.shares;
+  vmShares = filter (s: s.target-vm != null) cfg.shares;
+  filterSharesByName = vm: filter (s: s.target-vm == vm) vmShares;
+  sharePaths = map (s: s.host-path) filterHostShares;
+  vmList = unique (map (s: s.target-vm) vmShares);
+  mergeAttrSets = attrSets: foldl' (a: b: recursiveUpdate a b) {} attrSets;
+
+  # Share configuration
+  shareConfigs = vm: {
+    microvm.shares = forEach (filterSharesByName vm) (
+      vmShare: {
+        tag = "${vmShare.tag}";
+        source = "${vmShare.host-path}";
+        mountPoint = "${
+          if (vmShare.tmp-path != null)
+          then vmShare.tmp-path
+          else vmShare.vm-path
+        }";
+        proto = "virtiofs";
+      }
+    );
+  };
+
+  # Service configuration
+  serviceConfigs = vm:
+    forEach (filterSharesByName vm) (vmShare: {
+      environment.systemPackages = optionals (vmShare.tmp-path != null) [pkgs.umount];
+      systemd.services."storage-prep-${vmShare.tag}" = let
+        # Service to set owner, permissions, and (optionally) copy files
+        prepStorage = pkgs.writeShellScriptBin "storage_prep" ''
+          set -xeuo pipefail
+          if [ ! -d "${vmShare.vm-path}" ]; then
+            mkdir -p ${vmShare.vm-path}
+          fi
+          ${optionalString (vmShare.tmp-path != null) "cp -r ${vmShare.tmp-path}/* ${vmShare.vm-path}"}
+          chown -R ${vmShare.target-owner}:${vmShare.target-group} ${vmShare.vm-path}
+          chmod -R ${vmShare.target-permissions} ${vmShare.vm-path}
+          ${optionalString (vmShare.tmp-path != null) "${pkgs.umount}/bin/umount ${vmShare.tmp-path}"}
```

Comment:
Is this not something that would need to run inside the VM? As far as I understand the tmp-path is not mounted on the host but it's virtiofsd exporting host path as a filesystem to the VM.

Reply:
Yes the service configs are passed to the VMs.

Reply:
ok. I now wrapped my hand around it.